server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java

```diff
@@ -0,0 +1,604 @@
+                .and(UserTablePermissionSpec.filterByExpirationStatus(params.getStatuses(), expireTimeThreshold));
+        return userTablePermissionRepository.findAll(spec).stream().map(
+                e -> {
+                    UserTablePermission permission = mapper.entityToModel(e);
+                    Date expireTime = permission.getExpireTime();
+                    if (expireTime.before(expireTimeThreshold)) {
+                        permission.setStatus(ExpirationStatusFilter.EXPIRED);
+                    } else if (expireTime.before(DateUtils.addDays(expireTimeThreshold, EXPIRING_DAYS))) {
+                        permission.setStatus(ExpirationStatusFilter.EXPIRING);
+                    } else {
+                        permission.setStatus(ExpirationStatusFilter.NOT_EXPIRED);
+                    }
+                    return permission;
+                }).collect(Collectors.toList());
+    }
+
+    @Transactional(rollbackFor = Exception.class)
+    @SkipAuthorize("odc internal usage")
+    public List<UserTablePermission> listWithoutPageByDataSourceId(@NotNull Long dataSourceId,
+            @NotNull QueryTablePermissionParams params) {
+        Date expiredTime = new Date(System.currentTimeMillis() - expiredRetentionTimeSeconds * 1000);
+        permissionService.deleteExpiredPermission(expiredTime);
+        Date expireTimeThreshold = TimeUtils.getStartOfDay(new Date());
+        Specification<UserTablePermissionEntity> spec = Specification
+                .where(UserTablePermissionSpec.dataSourceId(dataSourceId))
+                .and(UserTablePermissionSpec.organizationIdEqual(authenticationFacade.currentOrganizationId()))
+                .and(UserTablePermissionSpec.userIdEqual(params.getUserId()))
+                .and(UserTablePermissionSpec.filterByExpirationStatus(params.getStatuses(), expireTimeThreshold));
+        return userTablePermissionRepository.findAll(spec).stream().map(
+                e -> {
+                    UserTablePermission permission = mapper.entityToModel(e);
+                    Date expireTime = permission.getExpireTime();
+                    if (expireTime.before(expireTimeThreshold)) {
+                        permission.setStatus(ExpirationStatusFilter.EXPIRED);
+                    } else if (expireTime.before(DateUtils.addDays(expireTimeThreshold, EXPIRING_DAYS))) {
+                        permission.setStatus(ExpirationStatusFilter.EXPIRING);
+                    } else {
+                        permission.setStatus(ExpirationStatusFilter.NOT_EXPIRED);
+                    }
+                    return permission;
+                }).collect(Collectors.toList());
+    }
+
+    @Transactional(rollbackFor = Exception.class)
+    @PreAuthenticate(hasAnyResourceRole = {"OWNER", "DBA"}, resourceType = "ODC_PROJECT", indexOfIdParam = 0)
+    public List<UserTablePermission> batchCreate(@NotNull Long projectId,
```

Comment:
Can we merge these two method `batchCreate` , `taskBatchCreate`?

Reply:
done